lib/SILOptimizer/Mandatory/DiagnoseInfiniteRecursion.cpp

```diff
@@ -0,0 +1,238 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements a diagnostic pass that detects deleterious forms of
+// recursive functions.
+//
+//===----------------------------------------------------------------------===//
+
+#define DEBUG_TYPE "infinite-recursion"
+#include "swift/AST/DiagnosticsSIL.h"
+#include "swift/AST/Expr.h"
+#include "swift/Parse/Lexer.h"
+#include "swift/SIL/CFG.h"
+#include "swift/SIL/SILArgument.h"
+#include "swift/SIL/SILInstruction.h"
+#include "swift/SILOptimizer/Analysis/PostOrderAnalysis.h"
+#include "swift/SILOptimizer/PassManager/Passes.h"
+#include "swift/SILOptimizer/PassManager/Transforms.h"
+#include "swift/SILOptimizer/Utils/Devirtualize.h"
+#include "llvm/ADT/PostOrderIterator.h"
+#include "llvm/ADT/iterator_range.h"
+#include "llvm/Support/Debug.h"
+
+using namespace swift;
+
+template<typename...T, typename...U>
+static void diagnose(ASTContext &Context, SourceLoc loc, Diag<T...> diag,
+                     U &&...args) {
+  Context.Diags.diagnose(loc,
+                         diag, std::forward<U>(args)...);
+}
+
+namespace {
+  class DiagnoseInfiniteRecursion : public SILFunctionTransform {
+  public:
+    DiagnoseInfiniteRecursion() {}
+
+  private:
+    static bool hasRecursiveCallInPath(SILBasicBlock &Block,
+                                       SILFunction *Target) {
+      // Process all instructions in the block to find applies that reference
+      // the parent function.  Also looks through vtables for statically
+      // dispatched (witness) methods.
+      auto *TargetModule = Target->getModule().getSwiftModule();
+      for (auto &I : Block) {
+        auto *AI = dyn_cast<ApplyInst>(&I);
+        if (AI && AI->getCalleeFunction() && AI->getCalleeFunction() == Target)
+          return true;
+
+        if (FullApplySite FAI = FullApplySite::isa(&I)) {
+          // Don't touch dynamic dispatch.
+          if (isa<ObjCMethodInst>(FAI.getCallee()))
+            continue;
+
+          auto &M = FAI.getModule();
+          if (auto *CMI = dyn_cast<ClassMethodInst>(FAI.getCallee())) {
+            auto ClassType = CMI->getOperand()->getType();
+
+            // FIXME: If we're not inside the module context of the method,
+            // we may have to deserialize vtables.  If the serialized tables
+            // are damaged, the pass will crash.
+            //
+            // Though, this has the added bonus of not looking into vtables
+            // outside the current module.  Because we're not doing IPA, let
+            // alone cross-module IPA, this is all well and good.
```

Comment:
Do we need to analyze `class_method` at all? That might not be an infinitely recursive call if the subclass overrides this method anyway.

(`witness_method` won't have that problem for the single-method case, though, because we don't have `super` calls.)

Reply:
> Do we need to analyze class_method at all?

See the `T` and `S` part of the regression test.  I would also say that even if the method is overriden, that as long as you can still instantiate something of the base class' type and call that function, you're still going to smash the stack so it's worthwhile to have the warning.

Reply:
Hm. There's no way to silence it, though. (Well, you can factor the recursive call out to a separate function, but that's kind of heavy.) I guess this won't come up much in practice, but I'm still kind of concerned...and it would dodge the whole problem of sometimes looking for a target class method.

Reply:
You can make it `dynamic` to silence the warning, which is kinda what you want anyways if you're implementing this kind of backwards-dispatch thing.

Reply:
I don't think it has anything to do with `dynamic`, but I guess it's pretty contrived for the base method to do some processing of its input and then call back to the subclass implementation on *all* paths. Even my attempts to make a silly example would be better off with a `fatalError`. Okay then.

Reply:
A "valid" example I thought of would be something trying to simulate SIMULA's `inner` keyword which would require pure virtual dispatch.  That's why I bring up `dynamic`

Reply:
Swift won't devirtualize unless it knows the type of `self`, so I don't think that *does* need `dynamic`.